gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcApiExceptionFactory.java

```diff
@@ -67,8 +87,76 @@ ApiException create(Throwable throwable) {
     }
   }
 
-  private ApiException create(Throwable throwable, Status.Code statusCode) {
+  private ApiException create(Throwable throwable, Status.Code statusCode, Metadata metadata) {
     boolean canRetry = retryableCodes.contains(GrpcStatusCode.grpcCodeToStatusCode(statusCode));
-    return ApiExceptionFactory.createException(throwable, GrpcStatusCode.of(statusCode), canRetry);
+    if (metadata == null) {
+      return ApiExceptionFactory.createException(
+          throwable, GrpcStatusCode.of(statusCode), canRetry);
+    }
+    byte[] bytes = metadata.get(Metadata.Key.of(ERROR_DETAIL_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+    if (bytes == null) {
+      return ApiExceptionFactory.createException(
+          throwable, GrpcStatusCode.of(statusCode), canRetry);
+    }
+
+    ErrorDetails.Builder errorDetailsBuilder = ErrorDetails.builder();
+    try {
+      com.google.rpc.Status status = com.google.rpc.Status.parseFrom(bytes);
+      List<String> detailsStringList = new ArrayList<>();
+      for (Any detail : status.getDetailsList()) {
+        Message unpack = unpack(detail);
+        if (unpack instanceof ErrorInfo) {
+          ErrorInfo errorInfo = (ErrorInfo) unpack;
+          errorDetailsBuilder.setDomain(errorInfo.getDomain());
+          errorDetailsBuilder.setReason(errorInfo.getReason());
+          errorDetailsBuilder.setErrorInfoMetadata(errorInfo.getMetadataMap());
+        }
+        detailsStringList.add("type: " + detail.getTypeUrl() + "\n" + unpack.toString());
+      }
+      errorDetailsBuilder.setDetailsStringList(detailsStringList);
```

Comment:
If the type is not `ErrorInfo`, would it even be relevant to the error details?

Reply:
Yes, all error message types in `error_details.proto` are relevant. `ErrorInfo` is the most important one but other messages may contain useful information as well, like `DebugInfo`.